### ISSUE TITLE
memory leak in function "ParserATNSimulator::execATNWithFullContext"

### ISSUE THREAD

**Issue by zhunki** [antlr/antlr4#2164]
when break out from the following loop:

  while (true) {
    reach = computeReachSet(previous, t, fullCtx);
    if (reach == nullptr) {
      // if any configs in previous dipped into outer context, that
      // means that input up to t actually finished entry rule
      // at least for LL decision. Full LL doesn't dip into outer
      // so don't need special case.
      // We will get an error no matter what so delay until after
      // decision; better error message. Also, no reachable target
      // ATN states in SLL implies LL will also get nowhere.
      // If conflict in states that dip out, choose min since we
      // will get error no matter what.
      NoViableAltException e = noViableAlt(input, outerContext, previous, startIndex);
      input->seek(startIndex);
      size_t alt = getSynValidOrSemInvalidAltThatFinishedDecisionEntryRule(previous, outerContext);
      if (alt != ATN::INVALID_ALT_NUMBER) {
        return alt;
      }
      throw e;
    }
    if (previous != s0) // Don't delete the start set.
        delete previous;
    previous = nullptr;

    std::vector<BitSet> altSubSets = PredictionModeClass::getConflictingAltSubsets(reach.get());
    reach->uniqueAlt = getUniqueAlt(reach.get());
    // unique prediction?
    if (reach->uniqueAlt != ATN::INVALID_ALT_NUMBER) {
      predictedAlt = reach->uniqueAlt;
      break;
    }
    if (_mode != PredictionMode::LL_EXACT_AMBIG_DETECTION) {
      predictedAlt = PredictionModeClass::resolvesToJustOneViableAlt(altSubSets);
      if (predictedAlt != ATN::INVALID_ALT_NUMBER) {
        break;
      }
    } else {
      // In exact ambiguity mode, we never try to terminate early.
      // Just keeps scarfing until we know what the conflict is
      if (PredictionModeClass::allSubsetsConflict(altSubSets) && PredictionModeClass::allSubsetsEqual(altSubSets)) {
        foundExactAmbig = true;
        predictedAlt = PredictionModeClass::getSingleViableAlt(altSubSets);
        break;
      }
      // else there are multiple non-conflicting subsets or
      // we're not sure what the ambiguity is yet.
      // So, keep going.
    }
    previous = reach.release();

    if (t != Token::EOF) {
      input->consume();
      t = input->LA(1);
    }
  }

the last reach is not freed.

the proposed patch(add a reach.reset(); before returns.):
  // If the configuration set uniquely predicts an alternative,
  // without conflict, then we know that it's a full LL decision
  // not SLL.
  if (reach->uniqueAlt != ATN::INVALID_ALT_NUMBER) {
    reportContextSensitivity(dfa, predictedAlt, reach.get(), startIndex, input->index()); 
    reach.reset();
    return predictedAlt;
  }

  // We do not check predicates here because we have checked them
  // on-the-fly when doing full context prediction.

  /*
   In non-exact ambiguity detection mode, we might	actually be able to
   detect an exact ambiguity, but I'm not going to spend the cycles
   needed to check. We only emit ambiguity warnings in exact ambiguity
   mode.

   For example, we might know that we have conflicting configurations.
   But, that does not mean that there is no way forward without a
   conflict. It's possible to have nonconflicting alt subsets as in:

   LL altSubSets=[{1, 2}, {1, 2}, {1}, {1, 2}]

   from

   [(17,1,[5 $]), (13,1,[5 10 $]), (21,1,[5 10 $]), (11,1,[$]),
   (13,2,[5 10 $]), (21,2,[5 10 $]), (11,2,[$])]

   In this case, (17,1,[5 $]) indicates there is some next sequence that
   would resolve this without conflict to alternative 1. Any other viable
   next sequence, however, is associated with a conflict.  We stop
   looking for input because no amount of further lookahead will alter
   the fact that we should predict alternative 1.  We just can't say for
   sure that there is an ambiguity without looking further.
   */
  reportAmbiguity(dfa, D, startIndex, input->index(), foundExactAmbig, reach->getAlts(), reach.get());
  reach.reset();
  return predictedAlt;

**Comment by niccroad** [comment]
Hi,
Can you clarify where you think the memory leak is? The pointer reach is of type std::unique_ptr<ATNConfigSet> so is deleted on return and calling reach.reset(); is spurious before any return statement. It will already be free'd when it goes out of scope.

http://en.cppreference.com/w/cpp/memory/unique_ptr

**Comment by mike-lischke** [comment]
@zhunki is right that there's a memory leak, but the solution is wrong. The leak is in the error case. If the there was at least one loop before the error case, `previous` contains the value from the previous `reach` set and must be freed now (just like it is done if there is no error. Unfortunately, `previous` is used in the error case, so we cannot just move the `delete` call to before the `if (reach == nullptr)` check.

**Comment by niccroad** [comment]
I think I get it. Sounds like this can be resolved if previous is a unique_ptr and ownership is transferred to the exception when its thrown (along with #2178).

**Comment by mike-lischke** [comment]
Unforunately not. `previous` initially holds `s0` and that must not be freed.

**Comment by garzon** [comment]
Maybe you can add a member in `ATNConfigSet` which is a pointer used to track the previous `reach`. Then the dtor of exception class could clean them up by following the linked list until `s0`.

**Comment by garzon** [comment]
I think, a better solution is, if the pointer should be deleted, duplicate the pointer as `std::shared_ptr` in a newly added member of `NoViableAltException`, otherwise do nothing, keep other logic the same.

**Comment by mike-lischke** [comment]
Yes, I thought about all that already, but then asked myself: is it really worth the effort to avoid a leak of one object in an error case where it doesn't matter anyway?

**Comment by garzon** [comment]
Maybe it is still worthy in some cases. In fact, I am using antlr in a project, parsing a lot of files with potential syntax errors.

```cpp
NoViableAltException e = noViableAlt(input, outerContext, previous, startIndex);
e->sharedPtrToDelete = previous;
```
A fix, like this, is enough in my case. But note that it is still problematic if a user duplicate the pointer from the exception and use it out of the scoop of the exception.

P.S. It is indicated by the tools provided in MSVC, that the return value of `getEpsilonTarget` seems remaining not being released after everything is destroyed. But I am still not sure where is the problem(just guessing, circular reference?), maybe you can have a look.

**Comment by mike-lischke** [comment]
I took back the half baken solution and instead added a new PR #2398.
